src/lib.rs

```diff
@@ -0,0 +1,111 @@
+extern crate libc;
+extern crate rand;
+
+use libc::c_void;
+use rand::Rng;
+use std::cmp;
+use std::env;
+use std::mem;
+use std::ptr;
+use std::sync::atomic::{self, AtomicUsize};
+
+const CANARY_SIZE: usize = mem::size_of::<usize>();
+
+static MEM_INIT: AtomicUsize = atomic::ATOMIC_USIZE_INIT;
+static CONF_ALLOC_EXTRA_MEM: AtomicUsize = atomic::ATOMIC_USIZE_INIT;
+
+pub extern fn libdiffuzz_init_config() {
+    if env::var_os("LIBDIFFUZZ_NONDETERMINISTIC").is_some() {
+        MEM_INIT.store(
+            rand::thread_rng().gen::<u8>().into(),
+            atomic::Ordering::Relaxed,
+        );
+    }
+    let alloc_extra_mem = env::var("LIBDIFFUZZ_ALLOCATE_EXTRA_MEMORY")
+        .ok()
+        .and_then(|x| x.parse().ok())
+        .unwrap_or(0);
+    CONF_ALLOC_EXTRA_MEM.store(alloc_extra_mem, atomic::Ordering::Relaxed);
+}
+
+#[link_section = ".ctors"]
+pub static CONSTRUCTOR: extern fn() = libdiffuzz_init_config;
+
+/// Gets then increments MEM_INIT
+fn get_mem_init() -> u8 {
+    MEM_INIT.fetch_add(1, atomic::Ordering::Relaxed) as u8
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn malloc(len: usize) -> *mut c_void {
+    let alloc_extra_mem = CONF_ALLOC_EXTRA_MEM.load(atomic::Ordering::Relaxed);
+    let full_len = match len.checked_add(CANARY_SIZE).and_then(|x| x.checked_add(alloc_extra_mem)) {
+        Some(x) => x,
+        None => return ptr::null_mut(),
+    };
+    let mut ptr = libc::mmap(
+        ptr::null_mut(),
+        full_len,
+        libc::PROT_READ | libc::PROT_WRITE,
+        libc::MAP_PRIVATE | libc::MAP_ANONYMOUS,
+        -1,
+        0,
+    );
+    // This is guaranteed to be aligned
+    *(ptr as *mut usize) = full_len;
+    ptr = ptr.offset(CANARY_SIZE as isize);
+    libc::memset(ptr, get_mem_init().into(), len + alloc_extra_mem);
+    ptr
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn calloc(n_items: usize, item_len: usize) -> *mut c_void {
+    let len = match n_items.checked_mul(item_len) {
+        Some(x) => x,
+        None => return ptr::null_mut(),
+    };
+    let alloc_extra_mem = CONF_ALLOC_EXTRA_MEM.load(atomic::Ordering::Relaxed);
+    let full_len = match len.checked_add(CANARY_SIZE).and_then(|x| x.checked_add(alloc_extra_mem)) {
+        Some(x) => x,
+        None => return ptr::null_mut(),
+    };
+    let mut ptr = libc::mmap(
+        ptr::null_mut(),
+        full_len,
+        libc::PROT_READ | libc::PROT_WRITE,
+        libc::MAP_PRIVATE | libc::MAP_ANONYMOUS,
+        -1,
+        0,
+    );
+    // This is guaranteed to be aligned
+    *(ptr as *mut usize) = full_len;
+    ptr = ptr.offset(CANARY_SIZE as isize);
+    libc::memset(ptr, 0, len);
+    libc::memset(
+        ptr.offset(len as isize),
+        get_mem_init().into(),
```

Comment:
`get_mem_init()` returns `u8`, but `memset()` accepts a `c_int`, which is usually much bigger and also signed. As far as I understand, if `get_mem_init()` returns 0xAB, then the memory will be clobbered with a repeating pattern of 000000AB, not ABABABAB as we'd like. I believe this was also an issue in the C version.

Do you see a way to achieve clobbering with ABABABAB without resorting to nightly-only API? The only way that springs to mind is to use `usize::from_bytes()`. There is probably a clever way to do that with bit shifts, but I don't see it.

Reply:
http://www.cplusplus.com/reference/cstring/memset/

> Value to be set. The value is passed as an int, but the function fills the block of memory using the unsigned char conversion of this value.

I don't believe that is the case.

Reply:
As I thought, this passes:

```rust
#[test]
fn test_memset() {
    let mut bytes = [0u8; 5];
    unsafe { libc::memset(bytes.as_mut_ptr() as _, 0xAB, bytes.len()) };
    assert_eq!(bytes, [0xAB; 5]);
}
```